packages/core/src/scuttle.js

```diff
@@ -0,0 +1,181 @@
+  function toRE(except) {
+    // turn scuttleGlobalThis.exceptions regexes strings to actual regexes
+    if (!except.startsWith('/')) {
+      return except
+    }
+    const parts = except.split('/')
+    const pattern = parts.slice(1, -1).join('/')
+    const flags = parts[parts.length - 1]
+    return new RegExp(pattern, flags)
+  }
+}
+
+/**
+ * Runs scuttling on the globalRef. Use applyDefaultScuttling for full scope of options.
+ *
+ * @param {GlobalRef} globalRef - Reference to the global object.
+ * @param {Array<string|RegExp>} extraPropsToAvoid - List of additional properties to exclude from scuttling beyond the default ones.
+ */
+function performScuttleGlobalThis(globalRef, extraPropsToAvoid = new Array()) {
+  const props = new Array()
+  getPrototypeChain(globalRef).forEach((proto) =>
+    props.push(...getOwnPropertyNames(proto))
+  )
+
+  // support LM,SES exported APIs and polyfills
+  const avoidForLavaMoatCompatibility = ['Compartment', 'Error', 'globalThis']
+  const propsToAvoid = new Set([
+    ...avoidForLavaMoatCompatibility,
+    ...extraPropsToAvoid,
+  ])
+
+  const obj = create(null)
+  for (const prop of props) {
+    // eslint-disable-next-line no-inner-declarations
+    function set() {
+      warn(
+        `LavaMoat - property "${prop}" of globalThis cannot be set under scuttling mode. ` +
+          'To learn more visit https://github.com/LavaMoat/LavaMoat/pull/360.'
+      )
+    }
+    // eslint-disable-next-line no-inner-declarations
+    function get() {
+      throw new Error(
+        `LavaMoat - property "${prop}" of globalThis is inaccessible under scuttling mode. ` +
+          'To learn more visit https://github.com/LavaMoat/LavaMoat/pull/360.'
+      )
+    }
+    if (shouldAvoidProp(propsToAvoid, prop)) {
+      continue
+    }
+    let desc = getOwnPropertyDescriptor(globalRef, prop)
+    if (desc?.configurable === true) {
+      desc = { configurable: false, set, get }
+    } else if (desc?.writable === true) {
+      const p = new Proxy(obj, { getPrototypeOf: get, get, set })
+      desc = { configurable: false, writable: false, value: p }
+    } else {
+      continue
+    }
+    defineProperty(globalRef, prop, desc)
+  }
+}
+
+/**
+ * @param {Set<string|RegExp>} propsToAvoid - List of properties to exclude from scuttling.
+ * @param {string} prop - Property to check.
+ * @returns {boolean} - Whether the property should be avoided or not.
+ */
+const shouldAvoidProp = (propsToAvoid, prop) =>
+  from(propsToAvoid).some(
+    (avoid) => avoid === prop || new RegExp(avoid).test(prop)
+  )
+
+/**
+ * @param {object} value - object to get the prototype chain from.
+ * @returns {Array<object>} - Prototype chain as an array.
+ */
+function getPrototypeChain(value) {
+  const protoChain = []
+  let current = value
+  while (current) {
+    if (typeof current !== 'object' && typeof current !== 'function') {
+      break
+    }
+    protoChain.push(current)
+    current = getPrototypeOf(current)
+  }
```

Comment:
```suggestion
  while (current && (typeof current === 'object' || typeof current === 'function')) {
    protoChain.push(current)
    current = getPrototypeOf(current)
  }
```

Reply:
Reason? Or just style?

Reply:
style, I guess.  I guess I like to avoid `break` statements if not absolutely needed

Reply:
I personally find it less convenient to both read and debug `while` loops with complicated conditions, I prefer test those within the loop itself (personally)

Reply:
functions and objects are always truthy, no? Could be simplified further:

```suggestion
  while (typeof current === 'object' || typeof current === 'function') {
    protoChain.push(current)
    current = getPrototypeOf(current)
  }
```

Reply:
@legobeat that's what I wrote lol